### DISCUSSION TITLE
Loading partial content that contains Alpine directives, via XHR

### DISCUSSION THREAD

**Discussion by Saeven** [alpinejs/alpine#2652] [1. Help] [answered]
Hi Everyone!

I'm coming from the Laminas/Zend Framework world and would like to make use of Alpine.  We've got a pretty big codebase right now that primarily runs Laminas+Boostrap+JQuery, but in new module releases, we want to experiment with Laminas+Tailwind+Alpine which provides a much better developer experience.

One issue I've encountered, occurs when loading partial content via ajax. 

To set the table, consider:
* a main template
* a partial template
* loading the partial into the main template via one of two methods

**layout.twig**
```
<div x-data="sampleModel()" @section="displaySection">
    <button @click="$dispatch('section', {type: 'foo'})">Foo</button>
    <div id="targetContainer"></div>
</div>


<script>
    function sampleModel(){
        return {
            displaySection: function(event) {
                // more on this below
                Utilities.load('/module/' + event.detail.type, {}, 'sectionArea');
            }
        }
    }
</script>
```

The partial that we are trying to load via AJAX is very simple:

```
<div x-data="foo()">
    <div x-text="bar"></div>
</div>

<script>
    function foo() {
        return {
            bar: 'this is alpine',
        }
    }
</script>
```

## ... the loading part

Utilities is a TS file that can be distilled to a very simple structure.  I have two versions of it.  One with jQuery that works as expected (the AlpineJS directives are functional after they are loaded), and a second that uses Fetch and fails.

Utilities.js with jQuery (works)
```
import $ = require("jquery");

export class Utilities {
    static load(route: string, data: object, target: string, showSpinner: boolean) {
        $('#' + target).load(route, data); // jQuery.load
    }
}
````

And a second which explodes when trying to use the Fetch API

Utilities.js, without jQuery
```
export class Utilities {
    static load(route: string, data: object, target: string, showSpinner: boolean) {
        const formData = new FormData();
        for (let [key, value] of Object.entries(data)) {
            formData.append(key, value);
        }

        return fetch(
            route, {
                method: 'POST',
                headers: {
                    'X-Requested-With': 'XMLHttpRequest',
                },
                body: formData
            })
            .then(response => response.text())
            .then(html => {
                document.getElementById(target).innerHTML = html;
            });
    }
}
```
The stack trace when using Fetch, looks like:
<img width="610" alt="Screen Shot 2022-02-10 at 1 15 41 PM" src="https://user-images.githubusercontent.com/887224/153470968-d07c294d-f258-4fa1-a10e-6c5c98c0c072.png">

Any tips on getting Alpine to render when Fetch is used?  I'd like to do away with jQuery.

**Comment by ekwoka** [comment, marked as answer]
Inline script tags do not run when inserted via innerhtml for security reasons. This is browser based.

To get around this, either make your component code bundled into global scoped JS with Alpine.data, or domParse the fetched content, pick out the script tags and append them to the head/body before appending the remaining DOM to wherever you need it.

Or, put the Data objects directly Into the `x-data` field. This might be the easiest with scoping.

**Reply by Saeven** [reply]
Thanks for taking the time to write @ekwoka.  The partials are dynamically rewritten by a multitude of conditions, so I wouldn't be able to jam it in a global scope.

Putting things directly into x-data will be a good workaround for simple cases, but a lot of these templates contain complex JS setups that also contain conditional script.

Walking out the script tags sounds like an interesting approach.  I'll explore that - thank you.

**Comment by Saeven** [comment]
I netted on this here which so far, satisfies all my uses cases:

 static load(route: string, data: object, target: string, showSpinner: boolean) {
        const formData = new FormData();
        for (let [key, value] of Object.entries(data)) {
            formData.append(key, value);
        }

        return fetch(
            route, {
                method: 'POST',
                headers: {
                    'X-Requested-With': 'XMLHttpRequest',
                },
                body: formData
            })
            .then(response => response.text())
            .then(html => {
                // clear out the target
                const targetElement = document.getElementById(target);
                this.removeAllChildNodes(targetElement);
                const nodes = this.htmlToElements(html);

                // rewrite scripts
                nodes.scripts.forEach((script: any) => {
                    let scriptElement = document.createElement('script');
                    scriptElement.type = 'text/javascript';
                    scriptElement.text = script.innerHTML;

                    // copy attributes over
                    let i = -1, attrs = script.attributes, attr;
                    while (++i > attrs.length) {
                        scriptElement.setAttribute((attr = attrs[i]).name, attr.value);
                    }

                    targetElement.appendChild(scriptElement);
                });

                nodes.nodes.forEach((element: Node) => {
                    targetElement.appendChild(element);
                });
            });
    }

    private static removeAllChildNodes(parent: Node): void {
        while (parent.firstChild) {
            parent.removeChild(parent.firstChild);
        }
    }

    private static htmlToElements(html: string): { scripts: Array<any>, nodes: Array<any> } {
        let template = document.createElement('template'),
            nodesArray = Array<Node>(),
            scriptsArray = Array<Node>();

        template.innerHTML = html;
        const nodes = template.content.childNodes;
        for (let i in nodes) {
            const examinedNode = nodes[i];
            if (examinedNode.nodeType === Node.ELEMENT_NODE) {
                if (examinedNode.nodeName === 'SCRIPT') {
                    scriptsArray.push(examinedNode);
                    continue;
                }
                nodesArray.push(examinedNode);
            }
        }

        return {nodes: nodesArray, scripts: scriptsArray};
    }

**Reply by Saeven** [reply]
I'll tighten the types up after I experiment a bit more.

**Reply by ekwoka** [reply]
@saeven You can simply use `querySelector` on your root dom node to get all the scripts in an array.

should be more like

```js
const scripts = template.content.querySelectorAll('script').map((node) => {
  const script = document.createElement('script');
  script.textContent = node.textContent;
  return script
})

document.body.append(...scripts)

```

**Comment by ProxiBlue** [comment]
FWIW, this answer in SO, solves this nicely.

https://stackoverflow.com/a/20584396/2028224

Once you inject the AJAX block via inner/outerHTML, you can target it using the answers example call

```nodeScriptReplace(document.getElementById('ajax-injected-block'));```
